server/src/main/java/org/opensearch/common/xcontent/support/XContentMapValues.java

```diff
@@ -215,18 +216,37 @@ private static Object extractValue(String[] pathElements, int index, Object curr
      * then {@code a.b} will be kept in the filtered map.
      */
     public static Map<String, Object> filter(Map<String, ?> map, String[] includes, String[] excludes) {
-        return filter(includes, excludes).apply(map);
+        return filter(includes, excludes, true).apply(map);
+    }
+
+    /**
+     * Only keep properties in {@code map} that match the {@code includes} but
+     * not the {@code excludes}. An empty list of includes is interpreted as a
+     * wildcard while an empty list of excludes does not match anything.
+     * <p>
+     * If a property matches both an include and an exclude, then the exclude
+     * wins.
+     * <p>
+     * If an object matches, then any of its sub properties are automatically
+     * considered as matching as well, both for includes and excludes.
+     * <p>
+     * Dots in field names are treated as sub objects. So for instance if a
+     * document contains {@code a.b} as a property and {@code a} is an include,
+     * then {@code a.b} will be kept in the filtered map.
+     */
+    public static Map<String, Object> filter(Map<String, ?> map, String[] includes, String[] excludes, boolean caseSensitive) {
+        return filter(includes, excludes, caseSensitive).apply(map);
     }
 
     /**
      * Returns a function that filters a document map based on the given include and exclude rules.
      * @see #filter(Map, String[], String[]) for details
      */
-    public static Function<Map<String, ?>, Map<String, Object>> filter(String[] includes, String[] excludes) {
+    public static Function<Map<String, ?>, Map<String, Object>> filter(String[] includes, String[] excludes, boolean caseSensitive) {
         if (hasNoWildcardsOrDots(includes) && hasNoWildcardsOrDots(excludes)) {
```

Comment:
Given this class is annotated as @opensearch.internal, we should be okay just adding this new parameter. Else, we probably need to maintain method with new and old signature?

Reply:
There are 20+ usages (mostly in tests) for the other signature so I opted to keep it as default behavior and add a second with a parameter. This same pattern is present in many places (i.e. Setting.listSetting) which has many different overrides. I can follow up this PR with another to remove the old signature if you think that would be beneficial.